src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/AccountsReferenceDateController.java

```diff
@@ -0,0 +1,164 @@
+import uk.gov.companieshouse.web.accounts.controller.ConditionalController;
+import uk.gov.companieshouse.web.accounts.exception.ServiceException;
+import uk.gov.companieshouse.web.accounts.model.smallfull.AccountsReferenceDate;
+import uk.gov.companieshouse.web.accounts.model.smallfull.AccountsReferenceDateQuestion;
+import uk.gov.companieshouse.web.accounts.model.state.CompanyAccountsDataState;
+import uk.gov.companieshouse.web.accounts.service.company.impl.CompanyServiceImpl;
+import uk.gov.companieshouse.web.accounts.service.navigation.NavigatorService;
+import uk.gov.companieshouse.web.accounts.service.smallfull.impl.SmallFullServiceImpl;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.validation.Valid;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@Controller
+@NextController(DirectorsReportQuestionController.class)
+@PreviousController(AccountsReferenceDateQuestionController.class)
+@RequestMapping("/company/{companyNumber}/transaction/{transactionId}/company-accounts/{companyAccountsId}/small-full/accounts-reference-date")
+public class AccountsReferenceDateController extends BaseController implements ConditionalController {
+
+    @Autowired
+    NavigatorService navigatorService;
+
+    @Autowired
+    CompanyServiceImpl companyService;
+
+    @Autowired
+    ApiClientService apiClientService;
+
+    @Autowired
+    private SmallFullServiceImpl smallFullService;
+
+    @Autowired
+    private HttpServletRequest request;
+
+    private static final String ACCOUNTS_REFERENCE_DATE = "accountsReferenceDate";
+
+
+    @GetMapping
+    public String getAccountsReferenceDateQuestion(@PathVariable String companyNumber,
+                                                   @PathVariable String transactionId,
+                                                   @PathVariable String companyAccountsId,
+                                                   Model model,
+                                                   HttpServletRequest request) {
+
+        try {
+            addBackPageAttributeToModel(model, companyNumber, transactionId, companyAccountsId);
+
+            AccountsReferenceDate accountsReferenceDate = new AccountsReferenceDate();
+
+            ApiClient apiClient = apiClientService.getApiClient();
+            CompanyProfileApi companyProfile = companyService.getCompanyProfile(companyNumber);
+            SmallFullApi smallFullAccounts = smallFullService.getSmallFullAccounts(apiClient, transactionId, companyAccountsId);
+
+            accountsReferenceDate.setPastDates(getPastValidDates(companyProfile.getAccounts().getNextAccounts().getPeriodEndOn()));
+            accountsReferenceDate.setFutureDates(getFutureValidDates(companyProfile.getAccounts().getNextAccounts().getPeriodEndOn()));
+
+            if(!companyProfile.getAccounts().getNextAccounts().getPeriodEndOn().equals(smallFullAccounts.getNextAccounts().getPeriodEndOn())) {
```

Comment:
Could we maybe abstract this into a private method? No fault of yours, it's just not the easiest to read!

Reply:
Implemented